crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -1462,3 +1462,79 @@ where
         state_provider.basic_account(address)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use reth_chain_state::ExecutedBlock;
+    use reth_primitives::{BlockNumHash, B256};
+    use reth_storage_api::BlockIdReader;
+    use reth_testing_utils::generators::{self, random_block_range};
+
+    use crate::{providers::BlockchainProvider2, test_utils::create_test_provider_factory};
+
+    #[test]
+    fn test_block_id_reader() -> eyre::Result<()> {
+        // Initialize random number generator and provider factory
+        let mut rng = generators::rng();
+        let factory = create_test_provider_factory();
+
+        // Generate 10 random blocks and split into database and in-memory blocks
+        let blocks = random_block_range(&mut rng, 0..=10, B256::ZERO, 0..1);
+        let (db_blocks, in_mem_blocks) = blocks.split_at(5);
+
+        // Insert first 5 blocks into the database
+        let provider_rw = factory.provider_rw()?;
+        for block in db_blocks {
+            provider_rw.insert_historical_block(
+                block.clone().seal_with_senders().expect("failed to seal block with senders"),
+            )?;
+        }
+        provider_rw.commit()?;
+
+        // Create a new provider and set the canonical head
+        let provider = BlockchainProvider2::new(factory)?;
+        provider
+            .canonical_in_memory_state
+            .set_canonical_head(in_mem_blocks.first().unwrap().header.clone());
+
+        // Set the pending block in memory
+        let pending_block = in_mem_blocks.last().unwrap();
+        provider.canonical_in_memory_state.set_pending_block(ExecutedBlock {
+            block: Arc::new(pending_block.clone()),
+            senders: Default::default(),
+            execution_output: Default::default(),
+            hashed_state: Default::default(),
+            trie: Default::default(),
+        });
```

Comment:
is there a need for all these setups? i don't think we query the database inside the `BlockIdReader` methods, so we can just set the canonical, pending, safe and finalized, and then call the provider methods

Reply:
- I've just removed the `set_canonical_head` instruction which was indeed useless for us here.
- However, maybe I missed something but I have the impression that we are obliged to insert at least one block so that the `BlockchainProvider2::new` that fetches the last header can work (otherwise seems we have a header not found error). And since we are inserting a block I thought we could insert 5 to have a slightly more realistic setup.

What do you think?

Reply:
makes sense. Let's wait until https://github.com/paradigmxyz/reth/pull/10376 is merged, it introduced a bunch of helpers for generating random blocks that we'll also need to use here.

Reply:
Done